### PR TITLE
Inline loops

### PR DESCRIPTION
**Pull request by kiip1**
### Description
This PR aims to implement inline versions of loops. Whilst allowing for easier prototyping with the effect command feature, it also makes the Skript language more rich and easier to type. Abuse of this feature gets prevented by denying nested inline loops; which isn't a perfect solution, but at the end of the day, it's up to the developer.

### Goals
- Allow for quicker prototyping with effect command.
- Allow substitution of simple loops into a single line.

### Non-Goals
- Be a replacement for loops.
- Allow for complex logic inside single line loops.
- Give leave to nested single line loops.

### Current State
At the moment, more testing is required before merging. There is also some dirty code that is surely improveable. Due to concern about the syntax and unintended side-effects in relation to addons and async effects, as well as allowing dirty code from developers, another idea is to limit this syntax to effect command context only, forcing normal scripts to be written with the more verbose while section.

### Patterns
__Inline While__
- `do [effect] <.+> [(1:((now|once) [and] [then]))] while <.+>`
  ```
  do add (random integer between 1 and 5) to {_amount} while chance of 50%
  broadcast {_amount}
  ```
__Inline Loop__
- `do [effect] <.+> %number% times`
  ```
  set {_amount} to 0
  do add 2 to {_amount} 5 times
  broadcast {_amount}
  ```
- `do [effect] <.+> for [(each|every) (element|item|entry) (in|of)] %objects%`
  ```
  do add uuid of loop-player to {_uuids::*} for every element in all players
  broadcast join {_uuids::*} by ", "
  ```

---
- #2652 (this is a general solution on the issue)
- #2653 (inline for `loop X times:`)
- #5098 (this could be used for the same)

### CODE REVIEW

File: src/main/java/ch/njol/skript/lang/parser/ParserInstance.java [hunk 1]
```diff
@@ -321,6 +324,40 @@ public final boolean isCurrentSection(Class<? extends TriggerSection>... section
...
+	public void pushLoop(Loopable loopable) {
+		currentLoops.add(loopable);
+	}
+	
+	/**
+	 * Removes the last loop from the current loops list.
+	 */
+	public void popLoop() {
+		currentLoops.remove(currentLoops.size() - 1);
+	}
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/lang/parser/ParserInstance.java:360]: Very unsure about this, wouldn't be needed if there was a way to get the current effect from an expression. Have I missed something?

File: src/main/java/ch/njol/skript/effects/EffLoop.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		String rawEffect = parseResult.regexes.get(0).group();
+		
+		usesTimes = matchedPattern == 0;
+		if (usesTimes) {
+			times = LiteralUtils.defendExpression(expressions[0]);
+			if (!LiteralUtils.canInitSafely(times)) {
+				Skript.error("Can't understand this loop: '" + parseResult.expr + "'");
+				return false;
+			}
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/effects/EffLoop.java:75]: Does it still need to be defended when a number is guaranteed?

File: src/main/java/ch/njol/skript/effects/EffLoop.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		try {
+			getParser().pushLoop(this);
+			effect = Effect.parse(rawEffect, "Can't understand this effect: " + rawEffect);
+		} finally {
+			getParser().popLoop();
+		}
+		if (effect instanceof EffLoop || effect instanceof EffWhile) {
+			Skript.error("Inline loop may not be nested inside another inline iterator.");
+			return false;
+		}
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/effects/EffLoop.java:93]: This code structure is dependant on the fact that getLoopedExpression() must be not-null before the Effect is parsed.

File: src/main/java/ch/njol/skript/lang/Loopable.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A simple interface to state a syntax element loopable, used by ExprLoopValue.
+ * Please note a while loop is not "loopable" since doesn't conform to this interface
+ * because it doesn't supply a looped expression nor currently iterated object.
+ */
+public interface Loopable {
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/lang/Loopable.java:30]: This could use a better name.

File: src/test/skript/tests/syntaxes/effects/EffLoop.sk [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	spawn a cow at spawn of "world"
+	set {_first} to last spawned cow
+	spawn a cow at spawn of "world"
+	set {_second} to last spawned cow
+	do add uuid of loop-entity to {_uuids::*} for every element in all cows
+	assert size of {_uuids::*} is 2 with "inline loop foreach failed"
+	loop {_uuids::*}:
+		assert loop-value is (uuid of {_first}) or (uuid of {_second}) with "inline loop foreach content failed"
+	delete {_first}
+	delete {_second}
```
Comment (Fusezion) [inline src/test/skript/tests/syntaxes/effects/EffLoop.sk:15]: Doing `delete {_var}` here shouldn't delete the entity, should change it to `delete entities`

File: src/main/java/ch/njol/skript/effects/EffLoop.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.stream.LongStream;
+
+@Name("Inline Loop")
+@Description("A shorthand effect for iterating over objects or repeating X times.")
+@Examples({"command /uuids:",
+	"/ttrigger:",
+	"\t\tdo add uuid of loop-player to {_uuids::*} for every element in all players",
+	"\t\tsend (join {_uuids::*} by \", \") to player"})
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/effects/EffLoop.java:48]: Needs another example for `do <effect> X times`

**Review by Fusezion — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffWhile.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+@Name("Inline While")
+@Description("A shorthand effect for executing an effect while a condition is met.")
+@Examples({"on join:",
+	"\tadd (random integer between 1 and 5) to {_amount} while chance of 50%",
+	"\tgive {_amount} of dirt to player"})
+@Since("INSERT VERSION")
+public class EffWhile extends Effect {
+	
+	static {
+		Skript.registerEffect(EffWhile.class, "[(1:do)] <.+> while <.+>");
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/effects/EffWhile.java:44]: Probably will drop support for `do while` since it's very confusing.

File: src/main/java/ch/njol/skript/effects/EffWhile.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+			ranDoWhile = true;
+			effect.setNext(this);
+			return effect;
+		} else if (condition.check(event)) {
+			debug(event, true);
+			effect.setNext(this);
+			return effect;
+		} else {
+			debug(event, false);
+			effect.setNext(getNext());
```
Comment (kiip1) [inline src/main/java/ch/njol/skript/effects/EffWhile.java:85]: Not needed?

File: src/main/java/ch/njol/skript/effects/EffWhile.java (hunk 7 above)
Comment (Fusezion) [reply src/main/java/ch/njol/skript/effects/EffWhile.java:44]: I believe keeping `do while` is a better thing as there's times I'd prefer to make sure I always add 1 to stuff this is better than  doing it once before the while

File: src/test/skript/tests/syntaxes/effects/EffLoop.sk (hunk 5 above)
Comment (kiip1) [reply src/test/skript/tests/syntaxes/effects/EffLoop.sk:15]: image
Should delete the entity

Comment (Fusezion) [reply src/test/skript/tests/syntaxes/effects/EffLoop.sk:15]: from testing in game I don't believe it does
image


it's been like this forever so unless it's a test run only thing it doesn't

Comment (Fusezion) [reply src/test/skript/tests/syntaxes/effects/EffLoop.sk:15]: there is a pr for it just not exactly like it
https://github.com/SkriptLang/Skript/pull/5198

File: src/main/java/ch/njol/skript/effects/EffLoop.java (hunk 2 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLoop.java:75]: Yes. Defending is when you provide an object as the type.

You can use the `*` character in your expression to force the expression to be a literal. `%*number%` You'd then have to test that it's the first matched pattern. This means that the user has to type an actual number (1, 2, 3...) and not a variable or an expression.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/effects/EffLoop.java:75]: You can also use regex `<.+>`

**Comment by TheLimeGlass** [comment]
This is going to have many conflicts with Ali's loop pull request at https://github.com/SkriptLang/Skript/pull/4595

File: src/main/java/ch/njol/skript/effects/EffLoop.java (hunk 2 above)
Comment (kiip1) [reply src/main/java/ch/njol/skript/effects/EffLoop.java:75]: Makes sense, I don't want a limitation to literals only here so will keep it as-is.

**Comment by UnderscoreTud** [comment]
I believe that inline while loops should not be a feature because they can easily cause Skript to crash. With a standard while loop, it is possible to prevent crashes by using the "wait 1 tick" effect, but this is not an option with inline while loops. In my opinion, the potential for crashing outweighs the convenience of having this feature. However, I do understand that some people may find it useful and appreciate its inclusion.

**Comment by kiip1** [comment]
> I believe that inline while loops should not be a feature because they can easily cause Skript to crash. With a standard while loop, it is possible to prevent crashes by using the "wait 1 tick" effect, but this is not an option with inline while loops. In my opinion, the potential for crashing outweighs the convenience of having this feature. However, I do understand that some people may find it useful and appreciate its inclusion.

Although infinite loops may be a problem, it practically won't make any difference; occlusion of `wait a tick` in while loops will cause a crash too. Inline loops may use some recognition for infinite loops e.g. it will error if `do send "Hello" to player while true` is used or it will stop execution & notify admins whenever its iteration count goes above a certain threshold. Besides that, inline loops should only be used for simple operations, in fact, it can only be used for simple operations due to it being limited to one line. I do understand it being a problem, but with the right built-in mitigations most common types of abuse can be prevented.

**Comment by APickledWalrus** [comment]
Thanks so much for this PR :)

I'll start off with some concerns I and Kenzie have about the feature.

The implementation itself is generally fine, although there are a few minor things I'll mention here for the sake of the conversation:
1. The changes to ParserInstance are too specific. Skript should not be aware of what a "loop" is, rather just aware of sections in general. However, I'm also not really sure what options there are for Skript to support something like this. I don't think there is currently a way to get the "parent" of an expression, but I'd certainly want a different approach if this is to be merged.

This PR is very similar to https://github.com/SkriptLang/Skript/pull/2653, which was closed after a long discussion. In the end, it came down to a few major issues:
1. Impact on parsing. Patterns like these can be rough, especially since `<.+>` is a valid option to start with. It can also lead to long lines which hurt both parsing and readability. That might not matter so much now, but I still think it should be a concern.
2. There are cases where we might have unpredictable behavior, whether it be with addon additions or something such as async effects.
3. Due to these concerns, I worry that this syntax sugar is not really worth it. By "playing it safe", we can force a better standard for writing scripts. I also think it comes off a little unnatural at times and can be hard to follow. Consider:
```
while the player is flying:
    send "Weeeeeeeee" to player
```
vs
```
send "Weeeeeeee" to the player while the player is flying
# If we want to reduce it
send "Weeeeeeee" while player is flying
```

Overall, I tend to think that we are better off forcing users to *use* sections rather than adding sugar for one-liners. I suppose it's useful in some languages, but I don't really see the benefit with something like Skript.

I of course welcome debate on this feature.

**Comment by kiip1** [comment]
After some discussion we ended up at the idea of limiting this syntax to effect command context only, forcing normal scripts to be written with the more verbose while section, but allowing for freedom during prototyping and quick operations on Skript variables. Before making this, more feedback on this would be appreciated.

**Comment by sovdeeth** [comment]
Do you still intend to work on this?

**Comment by kiip1** [comment]
@sovdeeth I currently don't have the time / willpower to work on this, if anyone wants to take it over feel free to do so.

**Comment by sovdeeth** [comment]
closing due to inactivity, though anyone who wishes is free to revive it!
